### PR TITLE
ENH: treat list of namedtuples like list of dict in DataFrame()

### LINKED ISSUE #27329: BUG: namedtuples fields not checked on DataFrame constructor
I personally find it very surprising that we assume that if the first element is a NamedTuple, that all elements of the list will be exactly the same NamedTuples:

```
In [29]: Record1 = namedtuple('Record1', ('b','a'))                                                                                                           

In [30]: Record2 = namedtuple('Record2', ('a','c'))                                                                                                           

In [31]: records = [Record1(1, 2), Record2(3, 4)]                                                                                                             

In [32]: pd.DataFrame(records)                                                                                                                                
Out[32]: 
   b  a
0  1  2
1  3  4
```

So the field names of the subsequent namedtuples are simply ignored. 

The case of all the same tuples might be the most typical case, but it doesn't seem very safe for a default behaviour to not check this.

### PR DESCRIPTION
**Pull request by ghost**
Following discussion in https://github.com/pandas-dev/pandas/pull/27309 about unifying "list of records" behavior across the board.

- [X] closes #27329
- [ ] tests added / passed 
- [X] whatsnew entry
- [ ] double check performance impact and update docs accordingly if needed.

### CODE REVIEW

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -235,6 +235,7 @@ Other enhancements
 - :func:`read_excel` can now use ``openpyxl`` to read Excel files via the ``engine='openpyxl'`` argument. This will become the default in a future release (:issue:`11499`)
 - :func:`pandas.io.excel.read_excel` supports reading OpenDocument tables. Specify ``engine='odf'`` to enable. Consult the :ref:`IO User Guide <io.ods>` for more details (:issue:`9070`)
 - :class:`Interval`, :class:`IntervalIndex`, and :class:`~arrays.IntervalArray` have gained an :attr:`~Interval.is_empty` attribute denoting if the given interval(s) are empty (:issue:`27219`)
+- :class:`DataFrame` now treats lists of typing.NameTuple equivalently to lists of nametuples. The behavior of the latter has changed in this release, please see the relevant section in "Breaking Changes".
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: remove this (and just use the section below)

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -803,6 +804,57 @@ order of the resulting DataFrame has changed compared to previous pandas verison
 
    pd.DataFrame(data)
 
+DataFrame constructor treats list of namedtuple/dict in the same way
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: move to 1.0

File: doc/source/whatsnew/v0.25.0.rst [hunk 3]
```diff
@@ -803,6 +804,57 @@ order of the resulting DataFrame has changed compared to previous pandas verison
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previously, only the first element in the list was checked and if it was a
+namedtuple, the field names of that single tuple were used as the column names.
+Subsequent tuples were assumed to be of the same type, and their values were
+looked up by position. As a consequence, if subsequent tuples of different types
+were included, any additional fields were ignored, and if similarly named fields
+appeard in a different order, alignment was not performed.
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: appeared

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -803,6 +804,57 @@ order of the resulting DataFrame has changed compared to previous pandas verison
...
    pd.DataFrame(data)
 
+DataFrame constructor treats list of namedtuple/dict in the same way
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previously, only the first element in the list was checked and if it was a
+namedtuple, the field names of that single tuple were used as the column names.
+Subsequent tuples were assumed to be of the same type, and their values were
+looked up by position. As a consequence, if subsequent tuples of different types
+were included, any additional fields were ignored, and if similarly named fields
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: need the issue number reference (and this PR number reference)

File: pandas/_libs/lib.pyx [hunk 5]
```diff
@@ -312,28 +312,53 @@ def fast_unique_multiple_list_gen(object gen, bint sort=True):
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
-def dicts_to_array(dicts: list, columns: list):
+def dicts_to_array(dicts: list, _columns : list):
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: don't rename this

File: pandas/tests/frame/test_constructors.py [hunk 6]
```diff
@@ -1336,8 +1336,36 @@ def test_constructor_list_of_namedtuples(self):
         tm.assert_frame_equal(result, expected)
 
         # with columns
-        expected = DataFrame({"y": [1, 2], "z": [3, 4]})
-        result = DataFrame(tuples, columns=["y", "z"])
+        # namedtuples now behave like records, so columns
```
Comment (jreback) [inline pandas/tests/frame/test_constructors.py]: can you make this into a new test

**Review by jreback — CHANGES_REQUESTED** [review]
at a glance looks ok, ideally can simplify some of the cython code (and add comments) as somewhat hard to follow

**Comment by pilkibun** [comment]
pushed fixed for all the reviews, and I'll work on making this more readable in the next few days.
Also, need to very the final performance impact, I'm not sure the ones quoted in the docs are still accurate (should be better).

**Comment by pilkibun** [comment]
II wanted to take care of this in thanks to @jorisvandenbossche for helping with the previous PR, but I just don't want to go through another pandas code review experience. I'm sorry, Joris.
